test/compilable/isZeroInit.d

```diff
@@ -0,0 +1,53 @@
+alias AliasSeq(T...) = T;
+
+static foreach (T; AliasSeq!(bool, byte, short, int, long,
+                             ubyte, ushort, uint, ulong,
+                             char, wchar, dchar,
+                             float, double, real))
+{
+    static assert(__traits(isZeroInit, T) == (T.init is T(0)));
+    static assert(__traits(isZeroInit, T[2]) == (T.init is T(0)));
+}
+
+static assert(__traits(isZeroInit, void)); // For initializing arrays of element type `void`.
+static assert(__traits(isZeroInit, void*));
+static assert(__traits(isZeroInit, float[]));
+static assert(__traits(isZeroInit, Object));
+class C1 : Object
+{
+    int x = 1;
+}
+static assert(__traits(isZeroInit, C1)); // An Object's fields are irrelevant.
+
+struct S1
+{
+    int[] a;
+    int b;
+}
+struct S2
+{
+    float f;
+}
+static assert(__traits(isZeroInit, S1));
+static assert(!__traits(isZeroInit, S2));
+
+struct S3
+{
+    S2 x = { 0 };
+    float f = 0;
+}
+struct S4
+{
+    int a = 1;
+}
+static assert(__traits(isZeroInit, S1));
+static assert(!__traits(isZeroInit, S2));
+
+void test()
+{
+    int a = 3;
+    float b = 0;
+    // Providing a specific variable rather than a type isn't allowed.
+    static assert(!__traits(compiles, __traits(isZeroInit, a)));
+    static assert(!__traits(compiles, __traits(isZeroInit, b)));
```

Comment:
We usually put failing tests in the `fail_compilation` directory which allows to verify the error message as well.